sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamZetaSqlCatalog.java

```diff
@@ -333,6 +380,42 @@ private void addUdfsFromSchema() {
     }
   }
 
+  private void validateScalarFunctionImpl(ScalarFunctionImpl scalarFunction) {
+    for (FunctionParameter parameter : scalarFunction.getParameters()) {
+      validateJavaUdfCalciteType(parameter.getType(typeFactory));
+    }
+    validateJavaUdfCalciteType(scalarFunction.getReturnType(typeFactory));
+  }
+
+  /**
+   * Throws {@link UnsupportedOperationException} if Calcite type is not supported in Java UDF.
+   * Supported types are a subset of the corresponding Calcite types supported by {@link
+   * BeamJavaUdfCalcRule}.
+   */
+  private void validateJavaUdfCalciteType(RelDataType type) {
+    switch (type.getSqlTypeName()) {
+      case BIGINT:
+      case DOUBLE:
+      case BOOLEAN:
+      case VARCHAR:
+      case VARBINARY:
+        // These types are supported.
+        break;
+      case DECIMAL:
+      case DATE:
+      case TIME:
+      case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+      case TIMESTAMP:
+      case ARRAY:
+      case ROW:
+        throw new UnsupportedOperationException(
+            "Calcite type not allowed in Java UDF: " + type.getSqlTypeName().getName());
```

Comment:
nit: Possibly make this 'ZetaSQL Java UDF'? (also possibly make this the default exception too?)

Reply:
Done.